[AArch64][BOLT] Ensure tentative code layout for cold BBs runs.

paschalis-mpeis opened the pull request:
When split functions is used, BOLT may skip tentative code layout estimation in some cases, like:
- when there is no profile data for some blocks (ie cold blocks)
- when there are cold functions in lite mode
- when skip functions is used
     
However, when rewriting the binary we still need to compute PC-relative
distances between hot and cold basic blocks. Without cold layout
estimation, BOLT uses '0x0' as the address of the first cold block,
leading to incorrect estimations of any PC-relative addresses.
 
This affects large binaries as the relaxStub method expands more
branches than necessary using the short-jump sequence, at it wrongly
believes it has exceeded the branch distance boundary.
 
This increases code size with both a larger and slower sequence; however,
performance regression is expected to be minimal since this only affects
any called cold code.
 
Example of such an unnecessary relaxation:
from:
```armasm
b       .Ltmp1234
```
 
to:
```armasm
adrp    x16, .Ltmp1234
add     x16, x16, :lo12:.Ltmp1234
br      x16
```

paschalis-mpeis committed:
BOLT may never run the tentative code layout for cold BBs.

However, it might later require the addresses of cold BBs to compute
PC-relative jumps between hot and cold fragments of a function. Since
the addresses were never estimated, it uses 0x0 for cold BBs, resulting
in incorrect estimated addresses for any instructions within them.

As a result, LongJump often expands a branch that would otherwise fit in
a single instruction into a short jump.

For example, `relaxStub` might expand from:
```armasm
b       .Ltmp1234
```

to:
```armasm
adrp    x16, .Ltmp1234
add     x16, x16, :lo12:.Ltmp1234
br      x16
```

While this expansion is not wrong, it is unnecessary. Moreover, in some
large binaries this expansion has lead to runtime crashes. This should
be investigated further, as such expansions should not cause crashes.

paschalis-mpeis committed:
[AArch64][BOLT] Ensure tentative code layout for cold BBs runs.

When split functions is used, BOLT may skip tentative code layout
estimation in some cases, like:
- when there is no profile data for some blocks (ie cold blocks)
- when there are cold functions in lite mode
- when skip functions is used

However, when rewriting the binary we still need to compute PC-relative
distances between hot and cold basic blocks. Without cold layout
estimation, BOLT uses '0x0' as the address of the first cold block,
leading to incorrect estimations of any PC-relative addresses.

This affects large binaries as the relaxStub method expands more
branches than necessary using the short-jump sequence, at it wrongly
believes it has exceeded the branch distance boundary.

This increases code size with a larger and slower sequence; however,
performance regression is expected to be minimal since this only affects
called cold code.

Example of such an unnecessary relaxation:
from:
```armasm
b       .Ltmp1234
```

to:
```armasm
adrp    x16, .Ltmp1234
add     x16, x16, :lo12:.Ltmp1234
br      x16
```

samolisov reviewed (commented):
@paschalis-mpeis Thank you for the patch. LGTM, but I worked with a rewritten tentative layout analysis and may not remember all the specific details of the mainline version. So, this will be great if you can wait for an approve for a more experienced bolt developer. Thank you.

samolisov commented:
(2) Relaxing shouldn't have caused a runtime crash

rafaelauler commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -382,6 +385,11 @@ uint64_t LongJmpPass::tentativeLayoutRelocMode(
     DotAddress += Func->estimateConstantIslandSize();
     ++CurrentIndex;
   }
+  if (!ColdLayoutDone) {
```
Instead of the if() on line 360 gating whether to skip the non-emitted function, if the goal is really to ensure that the layout for the cold part of split functions is done, why not complete the cold layout here on line 388:

```
  if (!ColdLayoutDone) {
    DotAddress =
          tentativeLayoutRelocColdPart(BC, SortedFunctions, DotAddress);
    if (opts::HotFunctionsAtEnd)
        DotAddress = alignTo(DotAddress, opts::AlignText);
  }
```

paschalis-mpeis replied:
I see your point.

What I had in mind when writing the patch was to introduce a minimal change in the logic.
I'm not sure if running `tentativeLayoutRelocColdPart` at 388 would require any of the remaining code between 367-385 to run as well.

I would want some feedback from @yota9  on this, to see how can we refactor/simplify the pass he originally authored.

yota9 replied:
Hello. I would take a look into it. But the pass belongs to @rafaelauler , I only made a few bugfixes here :)

paschalis-mpeis replied:
Thanks @yota9 .

BTW, I've just released a reproducer on the bug I describe earlier on this PR (see [point (2)](https://github.com/llvm/llvm-project/pull/96609#issuecomment-2188867516) in my initial comment)

**The issue is:**
- #99848

A possible lead could be the FixRelaxations pass. Any feedback would be appreciated (cc: @yota9 @rafaelauler  @samolisov ).

yota9 replied:
After checking out the review again I agree with @rafaelauler. I would rather create lamda with 
`
      DotAddress =
          tentativeLayoutRelocColdPart(BC, SortedFunctions, DotAddress);
      ColdLayoutDone = true;
      if (opts::HotFunctionsAtEnd)
        DotAddress = alignTo(DotAddress, opts::AlignText);
`

And run it under previous condition and under only `if (!ColdLayoutDone)`  after the loop.

paschalis-mpeis replied:
Done, thanks!

yota9 commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -354,9 +354,12 @@ uint64_t LongJmpPass::tentativeLayoutRelocMode(
       HotAddresses[Func] = Func->getAddress();
-      continue;
+      // Don't perform any tentative address estimation of a function's cold
+      // layout if it won't be emitted, unless we are ignoring a large number of
```
I think If we won't skip working on these functions we would end up with setting new address on 376 line which would be the wrong address for this function

paschalis-mpeis replied:
Thanks @yota9 for your review. This should be addressed now. Can you please confirm?

yota9 commented on bolt/test/AArch64/split-funcs-lite.test:
```diff
@@ -0,0 +1,14 @@
+RUN: %clang %cflags %p/../Inputs/asm_main.c -Wl,-q -o %t
+
+RUN: llvm-bolt %t -o %t.bolt -lite=1 -split-functions -split-all-cold \
+RUN: --skip-funcs="_init,_start,call_weak_fn/1,deregister_tm_clones/1,register_tm_clones/1,__do_global_dtors_aux/1,frame_dummy/1,main,foo,_fini" 2>&1 | FileCheck %s
```
Since we're using nostdlib flag most of these symbols won't appear in the binary except foo and main.

yota9 commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -324,7 +325,6 @@ uint64_t LongJmpPass::tentativeLayoutRelocColdPart(
     uint64_t DotAddress) {
-
   // Compute hot cold frontier
   uint32_t LastHotIndex = -1u;
```
Also while you here please make LastHotIndex variable int64_t . Thank you!

paschalis-mpeis committed:
Addressing reviewers.

paschalis-mpeis commented:
Thanks all for your reviews. I've introduced a lambda which I call after the loop as suggested.

I force-pushed as I rebased to recent main.
I've also adjusted the initial 'showcase commit' (08dd00fd7e25) to reflect the changes.

paschalis-mpeis commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -382,6 +385,11 @@ uint64_t LongJmpPass::tentativeLayoutRelocMode(
     DotAddress += Func->estimateConstantIslandSize();
     ++CurrentIndex;
   }
+  if (!ColdLayoutDone) {
```
Hi @samolisov,

Thanks for your review. Now this is simplified and the cold layout always runs.

yota9 commented on bolt/lib/Passes/LongJmp.cpp:
```diff
@@ -382,6 +382,13 @@ uint64_t LongJmpPass::tentativeLayoutRelocMode(
+  // Ensure that tentative code layout always runs for cold blocks.
+  if (!ColdLayoutDone)
+    runColdLayout();
+  assert(ColdLayoutDone &&
```
It seems this assert is redundant and could not be raised by anything, I suggest to remove it

paschalis-mpeis replied:
True. I kept it just for testing but TBH I also did not like it.

This is now gone. I'm using a mock profile to trigger this, and I check whether any cold tentative estimation happened.

I've also updated the [description of this PR](https://github.com/llvm/llvm-project/pull/96609#issue-2372128097).

paschalis-mpeis committed:
Addressing reviewers - 02

yota9 commented on bolt/test/AArch64/split-funcs-lite.s:
```diff
@@ -0,0 +1,27 @@
+// RUN: %clang %cflags -o %t %s
+// RUN: %clang %s %cflags -Wl,-q -o %t
+// RUN: link_fdata --no-lbr %s %t %t.fdata
+// RUN: llvm-bolt %t -o %t.bolt --data %t.fdata -split-functions  \
```
Please remove extra space at the end

yota9 commented on bolt/test/AArch64/split-funcs-lite.s:
```diff
@@ -0,0 +1,27 @@
+// RUN: %clang %s %cflags -Wl,-q -o %t
+// RUN: link_fdata --no-lbr %s %t %t.fdata
+// RUN: llvm-bolt %t -o %t.bolt --data %t.fdata -split-functions  \
+// RUN: -debug 2>&1 | FileCheck %s
```
And add 2 spaces after :

paschalis-mpeis replied:
Done in both places, thanks.

yota9 reviewed (approved):
Thank you for the commit. Please fix indentation a bit a let's give 1-2 days for Meta team to review

paschalis-mpeis committed:
Addressing reviewers (3): fix indentation

paschalis-mpeis commented:
If no objections I will be merging this tomorrow.

maksfb reviewed (changes requested):
The change looks good to me, but I believe the test case will fail on non-debug builds because `LLVM_DEBUG` will be a no-op and the expected string wouldn't be printed.

maksfb commented:
@paschalis-mpeis, could you try the test from https://github.com/llvm/llvm-project/pull/112110 without `--compact-code-model`? It fails in JITLink (without the flag), but this PR supposedly fixes it.

paschalis-mpeis committed:
Addressed reviewers (4)

Changed comment style on the test.

paschalis-mpeis commented:
> the change looks good to me, but I believe the test case will fail on non-debug builds because LLVM_DEBUG will be a no-op and the expected string wouldn't be printed.

Thanks @maksfb. The test should now run only when compiled with assertions.

paschalis-mpeis commented:
@maksfb I played around with the test of #112110, and below are my findings.


In summary, with random splitting alone, no stubs are inserted, and no relocation is needed or performed. So as expected, the JITLink crash you observe does not go away.

It's only when I attach profile data that parts of LongJmp are engaged (incl. my patch). The JITLink crash is then is gone, but not due to my patch. In this simple/small test even with a wrong cold layout BOLT is not relaxing any branches. There is an example at the very end.

# Details
The current patch (#96609) avoids relaxing (with `relaxStubToShortJmp`) any branches that tentative code layout predicts they are not needed.

When trying the test of #112110, the flags `--split-functions --split-strategy=randomN` do not trigger `relaxStub`. Therefore, my proposed change does not run, no stubs are [added](https://github.com/llvm/llvm-project/blob/d03fa48eb1a8866eda24d8bed9165a9e28c0a3ce/bolt/lib/Passes/LongJmp.cpp#L606), and no relaxation is [needed](https://github.com/llvm/llvm-project/blob/d03fa48eb1a8866eda24d8bed9165a9e28c0a3ce/bolt/lib/Passes/LongJmp.cpp#L619).
**Was something different expected and could it be an issue with the random splitting strategy?** 🤔 

BTW, this was the reason for not using random splitting in my test, but instead a [profile](https://github.com/llvm/llvm-project/blob/d03fa48eb1a8866eda24d8bed9165a9e28c0a3ce/bolt/test/AArch64/split-funcs-lite.s#L8) with [internal labels](https://github.com/llvm/llvm-project/blob/d03fa48eb1a8866eda24d8bed9165a9e28c0a3ce/bolt/test/AArch64/split-funcs-lite.s#L21).

---
### Running test of #112110
When running the test (just w/o the compact-flag of course), the [fix-point loop](https://github.com/llvm/llvm-project/blob/d03fa48eb1a8866eda24d8bed9165a9e28c0a3ce/bolt/lib/Passes/LongJmp.cpp#L642) does a single iteration on `runOnFunctions` for both `_start`, and `large_function`, which results to no stub insertions and no relaxation:
> Inserted 0 stubs in the hot area and 0 stubs in the cold area. Shared 0 times, iterated 1 times.

As expected, the JITLink crash appears, either with or without this patch:
> BOLT-ERROR: JITLink failed: In graph in-memory object file, section .text: relocation target .text.cold + 0x4 at address 0x724fc0 is out of range of CondBranch19PCRel fixup at 0x600004 (_start, 0x600000 + 0x4)

### What is the intention of the #112110 test?
Using your patch, in the bolted binary we get something like:
```armasm
Disassembly of section .text:

0000000000600000 <_start>:
  600000: cmp	x1, #1
  600004: b.hi	0x60000c <_start+0xc>
  600008: b	0x724fc0 <_start.cold.0>
  60000c: b	0x724fc4 <_start.cold.0+0x4>

0000000000600010 <large_function>:
  600010: nop
...
  724f8c: nop
  724f90: ret

Disassembly of section .text.cold:

0000000000724fc0 <_start.cold.0>:
  724fc0: bl	0x600010 <large_function>
  724fc4: ret
```

And this aligns with your comments in the test. So we have:
- in the hot section: the main fragment `_start`  and the `large_function`
- in the cold section: the non-main fragment of `_start`

With the surrounding code as is, I only achieve this kind of splitting when I change your test to attach some profile data on the `_start` entry-point and the `large_function`. In that case, the **JITLink error is gone** and the resulting code is:
```armasm

Disassembly of section .text:

0000000000600000 <_start>:
  600000: cmp	x1, #0x1
  600004: b.ls	0x60000c <_start+0xc>
  600008: b	0x724fc4 <_start.cold.0+0x4>
  60000c: b	0x724fc0 <_start.cold.0>

0000000000600010 <large_function>:
  600010: nop
...
  724f8c: nop
  724f90: ret

Disassembly of section .text.cold:

0000000000724fc0 <_start.cold.0>:
  724fc0: bl	0x600010 <large_function>
  724fc4: ret
```

Note that the branching condition in the above snippet is switched to `b.ls` when compared to the original test. That must be due to the profile data I used that indicate a call to `large_function` is the more likely branch target.

So when profile data are attached, wider code of LongJmp runs (incl. my change), a couple of stubs are inserted, and the JITLink crassh is gone. But as far as my patch is concerned, in this simple/small example it has no effect.

For example, without my patch we'll get addresses (1):
- [SrcAddr](https://github.com/llvm/llvm-project/blob/d03fa48eb1a8866eda24d8bed9165a9e28c0a3ce/bolt/lib/Passes/LongJmp.cpp#L484): `0x600008`                                                         
- [TgtAddr](https://github.com/llvm/llvm-project/blob/d03fa48eb1a8866eda24d8bed9165a9e28c0a3ce/bolt/lib/Passes/LongJmp.cpp#L483): `0x4 `


and with my patch we'll get (2):
- SrcAddr: `0x600008`
- TgtAddr: `0x724fc4`

(1) is within bounds, so correcting the cold blocks estimated address has no effect.

maksfb reviewed (approved):
Thanks for the explanation, @paschalis-mpeis.
